// convex/src/server/impl/storage_impl.ts
import {
  FileMetadata,
  StorageHttpWriter,
  StorageId,
  StorageReader,
  StorageWriter,
} from "../storage.js";
import { performAsyncSyscall, performJsSyscall } from "./syscall.js";
import { validateArg } from "./validate.js";

export function setupStorageReader(): StorageReader {
  return {
    getUrl: async (storageId: StorageId) => {
      validateArg(storageId, 1, "getUrl", "storageId");
      return await performAsyncSyscall("storageGetUrl", { storageId });
    },
    getMetadata: async (storageId: StorageId): Promise<FileMetadata> => {
      return await performAsyncSyscall("storageGetMetadata", { storageId });
    },
  };
}

export function setupStorageWriter(): StorageWriter {
  const reader = setupStorageReader();
  return {
    generateUploadUrl: async () => {
      return await performAsyncSyscall("storageGenerateUploadUrl", {});
    },
    delete: async (storageId: StorageId) => {
      await performAsyncSyscall("storageDelete", { storageId });
    },
    getUrl: reader.getUrl,
    getMetadata: reader.getMetadata,
  };
}

export function setupStorageHttpWriter(): StorageHttpWriter {
  const writer = setupStorageWriter();
  return {
    ...writer,
    store: async (request: Request) => {
      return await performJsSyscall("storage/storeFile", {
        request,
      });
    },
    get: async (storageId: StorageId) => {
      return await performJsSyscall("storage/getFile", {
        storageId,
      });
    },
  };
}
